python/GangaDirac/Lib/Utilities/DiracUtilities.py

```diff
@@ -206,20 +206,42 @@ def execute(command,
     if python_setup == '':
         python_setup = getDiracCommandIncludes()
 
-    _checkProxy()
+    # We're about to perform an expensive operation so being safe before we run it shouldn't cost too much
+    _checkProxy(force = True, renew = renew)
 
     #logger.info("Executing command:\n'%s'" % str(command))
     #logger.debug("python_setup:\n'%s'" % str(python_setup))
     #logger.debug("eval_includes:\n'%s'" % str(eval_includes))
 
+    if cwd is None:
+        # We can in all likelyhood be in a temp folder on a shared (SLOW) filesystem
+        # If we are we do NOT want to execute commands which will involve any I/O on the system that isn't needed
+        cwd_ = tempfile.mkdtemp()
+    else:
+        # We know were whe want to run, lets just run there
+        cwd_ = cwd
+
+    global last_modified_valid
+    if not last_modified_valid:
+        return None
+
+    return_code = 0
     returnable = gexecute.execute(command,
                                   timeout=timeout,
                                   env=env,
-                                  cwd=cwd,
+                                  cwd=cwd_,
                                   shell=shell,
                                   python_setup=python_setup,
                                   eval_includes=eval_includes,
-                                  update_env=update_env)
+                                  update_env=update_env,
+                                  return_code=return_code)
+
+    ## Command failed, lets check to see if it was due to the proxy or not
+    if return_code:
+        last_modified_valid = False
```

Comment:
Not in an easy way. There are a couple of problems:

1. We run the passed command in a `bash "..."` subshell and I'm not sure if error codes are propagated out correctly.
2. If `update_env` is used, we are running a number of processes chained together and only the last's rc will be caught.
3. If running a Python command, I've no idea how much we can rely on sensible return codes (this one if probably ok though).

Once those potential problems are addressed then the only way to get the information out is to pass it as a structured return value (a `dict` or a `tuple` or something) or make the `return_value` object some mutable compound type. Alternatively, raising an exception would make sense I think so that it is always noticed if a command fails.